src/backend/storage/ipc/temp_tables_limit.c

```diff
@@ -0,0 +1,268 @@
+{
+	return sizeof(pg_atomic_uint64);
+}
+
+void
+TempTablesLimitShmemInit(void)
+{
+	temp_tables_limit_value = (volatile pg_atomic_uint64 *) ShmemAlloc(sizeof(pg_atomic_uint64));
+	pg_atomic_init_u64((pg_atomic_uint64 *) temp_tables_limit_value, 0);
+}
+
+/* ao */
+
+void
+BufferedAppendWritePreHook(BufferedAppend *bufferedAppend)
+{
+	TempTablesLimitChecks();
+
+	if (RelFileNodeBackendIsTemp(bufferedAppend->relFileNode))
+	{
+		prevFileLen = bufferedAppend->largeWritePosition;
+
+		if (bufferedAppend->largeWriteLen > 0 &&
+			temp_tables_limit_value->value + bufferedAppend->largeWriteLen > TempTablesLimitToBytes())
+			elog(ERROR, "Temp tables quota exceeded");
+	}
+}
+
+void
+BufferedAppendWritePostHook(BufferedAppend *bufferedAppend)
+{
+	TempTablesLimitChecks();
+
+	if (RelFileNodeBackendIsTemp(bufferedAppend->relFileNode))
+	{
+		Assert(prevFileLen >= 0);
+		pg_atomic_add_fetch_u64(temp_tables_limit_value, bufferedAppend->largeWriteLen);
+	}
+
+	prevFileLen = -1;
+}
+
+void
+TruncateAOSegmentFilePreHook(Relation rel, File fd)
+{
+	TempTablesLimitChecks();
+
+	if (rel->rd_islocaltemp)
+	{
+		prevFileLen = FileDiskSize(fd);
+	}
+}
+
+void
+TruncateAOSegmentFilePostHook(Relation rel, int64 offset)
+{
+	int64 bytesToTruncate;
+
+	TempTablesLimitChecks();
+
+	if (rel->rd_islocaltemp)
+	{
+		Assert(prevFileLen >= 0);
+
+		bytesToTruncate = prevFileLen - offset;
+		pg_atomic_sub_fetch_u64(temp_tables_limit_value, bytesToTruncate);
+	}
+
+	prevFileLen = -1;
+}
+
+/* heap */
+
+void
+mdextend_pre_hook(SMgrRelation reln, ForkNumber forknum, int64 size)
+{
+	TempTablesLimitChecks();
+
+	if (SmgrIsTemp(reln))
+		if (temp_tables_limit_value->value + size > TempTablesLimitToBytes()
+			&& forknum == MAIN_FORKNUM) /* visibility map can be extended during vacuum */
+			elog(ERROR, "Temp tables quota exceeded");
+}
+
+void
+mdextend_post_hook(SMgrRelation reln, int64 size)
+{
+	TempTablesLimitChecks();
+
+	if (SmgrIsTemp(reln))
+		pg_atomic_add_fetch_u64(temp_tables_limit_value, size);
+}
+
+void
+mdunlinkfork_pre_hook(RelFileNodeBackend rnode, ForkNumber forkNum)
+{
+	struct stat buf;
+	char *path;
+	int fd;
+
+	TempTablesLimitChecks();
+
+	if (RelFileNodeBackendIsTemp(rnode))
+	{
+		path = relpath(rnode, forkNum);
+		fd = OpenTransientFile((char *) path, O_RDWR | PG_BINARY, 0);
+		if (fstat(fd, &buf) == 0)
+		{
+			prevFileLen = buf.st_size;
+			CloseTransientFile(fd);
+		}
+		else
+			fileSkip = true; // file didn't exist, skip it
+	}
+}
+
+void
+mdunlinkfork_post_hook(RelFileNodeBackend rnode)
+{
+	TempTablesLimitChecks();
+
+	if (RelFileNodeBackendIsTemp(rnode) && !fileSkip)
+	{
+		Assert(prevFileLen >= 0);
+
+		pg_atomic_sub_fetch_u64(temp_tables_limit_value, prevFileLen);
+	}
+
+	prevFileLen = -1;
+	fileSkip = false;
+}
+
+void
+mdunlinkforksegment_pre_hook(RelFileNodeBackend rnode, char *segpath)
+{
+	struct stat buf;
+	char fullPath[MAXPGPATH];
+	size_t strLen = strnlen(data_directory, MAXPGPATH) + strnlen(segpath, MAXPGPATH) + 1;
```

Comment:
if strlen(data_directory) > MAXPGPATH and strlen(segpath) > MAXPGPATH then strLen > 2 * MAXPGPATH !

Reply:
this line is redundant